src/shadowbox/server/manager_metrics.ts

```diff
@@ -13,23 +13,23 @@
 // limitations under the License.
 
 import {PrometheusClient} from '../infrastructure/prometheus_scraper';
-import {DataUsageByUser} from '../model/metrics';
+import {DataUsageByUser, DataUsageTimeframe} from '../model/metrics';
 
 export interface ManagerMetrics {
-  get30DayByteTransfer(): Promise<DataUsageByUser>;
+  getOutboundByteTransfer(timeframe: DataUsageTimeframe): Promise<DataUsageByUser>;
 }
 
 // Reads manager metrics from a Prometheus instance.
 export class PrometheusManagerMetrics implements ManagerMetrics {
   constructor(private prometheusClient: PrometheusClient) {}
 
-  async get30DayByteTransfer(): Promise<DataUsageByUser> {
+  async getOutboundByteTransfer(timeframe: DataUsageTimeframe): Promise<DataUsageByUser> {
     // TODO(fortuna): Consider pre-computing this to save server's CPU.
```

Comment:
We populate `AccessKey.dataUsage` through `getOutboundByteTransfer`. I removed `ServerAccessKeyRepository.getDataUsage` because, with the shared timeframe, there is no need to compute individual access key usage.

Reply:
OK, but `ShadowsocksManagerService.getDataUsage` still calls `this.managerMetrics.getOutboundByteTransfer`, which is a full recomputation from Prometheus.  Now that we have hourly recomputations for all access keys, `getDataUsage` could return those cached values instead of forcing a recalculation.

We don't have to do this now, but I think the TODO here has become confusing because half of it is now implemented, and the other half wouldn't be implemented at this point in the code.

Reply:
Oh I like that. Here is a possible design:

- PrometheusManagerMetrics runs the polling loop to update the usage. We should do it more often. Maybe once a minute? What do we use in the manager polling? We would need to check the CPU.
- The bytes transferred API doesn't hit prometheus. It only returns the currently cached values.
- PrometheusManagerMetrics exposes a "metric update" event others can subscribe
- We use the event to enforce the data limits
- We use the event in the data limit tests tests (which can actually inject a fake ManagerMetrics).

This way:
- the quota enforcement is really driven by how fast we update the usage
- The manager can hit the api as often as it wants and it won't cause any more load.
- Later we may even change the manager to an event-based API that keeps an HTTPS connection waiting for the next metric update.

Reply:
BTW, the TODO was about materializing the query on Prometheus, but it's effectively the same.
We can't actually do it on Prometheus now though, since the timeframe can change (it was fixed before)